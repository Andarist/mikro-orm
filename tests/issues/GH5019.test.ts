import {
  MikroORM,
  Entity,
  PrimaryKey,
  Property,
  SimpleLogger,
  Collection,
  Ref,
  ManyToMany,
} from '@mikro-orm/sqlite';
import { mockLogger } from '../helpers';

@Entity()
class Book {

  @PrimaryKey()
  id!: number;

  @Property()
  title: string;

  @ManyToMany(() => BookTag)
  tags = new Collection<BookTag>(this);

  constructor({ id, title }: { id?: number; title: string }) {
    if (id) {
      this.id = id;
    }
    this.title = title;
  }

}

@Entity()
export class BookTag {

  @PrimaryKey()
  id!: number;

  @Property()
  name: string;

  @ManyToMany(() => Book, book => book.tags)
  books = new Collection<Book>(this);

  constructor({ id, name, books }: { id?: number; name: string; books: (Book | Ref<Book>)[] }) {
    if (id) {
      this.id = id;
    }
    this.name = name;
    this.books.add(books);
  }

}
describe('GH #5019', () => {
  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Book, BookTag],
      dbName: ':memory:',
      loggerFactory: options => new SimpleLogger(options),
    });
    await orm.schema.refreshDatabase();

    const books = [
      new Book({ id: 1, title: 'One' }),
      new Book({ id: 2, title: 'Two' }),
      new Book({ id: 3, title: 'Three' }),
      new Book({ id: 4, title: 'Four' }),
      new Book({ id: 5, title: 'Five' }),
      new Book({ id: 6, title: 'Six' }),
    ];
    orm.em.persist(books);

    const tags = [
      new BookTag({ id: 1, name: '1', books: [books[0]] }),
      new BookTag({ id: 2, name: '2', books: [books[0], books[3]] }),
      new BookTag({ id: 3, name: '3', books: [books[1], books[2], books[3]] }),
    ];
    orm.em.persist(tags);

    await orm.em.flush();
    orm.em.clear();
  });

  afterAll(() => orm.close(true));

  test('query 1', async () => {
    const mock = mockLogger(orm);
    const tags1 = await orm.em.fork().find(BookTag, { books: 1 }, { populate: ['books:ref'] });
    expect(mock.mock.calls).toMatchSnapshot(); // FIXME: exact query might vary once fixed
  });

  test('query 2', async () => {
    const mock = mockLogger(orm);
    const tags2 = await orm.em.fork().find(BookTag, { books: 1 }, { populate: ['books'] });
    expect(mock.mock.calls).toMatchSnapshot(); // FIXME: exact query might vary once fixed
  });

  test('query 3', async () => {
    const mock = mockLogger(orm);
    const tags3 = await orm.em.fork().find(BookTag, { id: 1 }, { populate: ['books'] });
    expect(mock.mock.calls).toMatchSnapshot();
  });
});
