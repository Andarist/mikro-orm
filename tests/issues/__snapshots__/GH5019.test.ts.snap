// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GH #5019 query 1 1`] = `
[
  [
    "[query] select \`b0\`.*, \`b1\`.\`book_tag_id\` as \`b1__book_tag_id\`, \`b1\`.\`book_id\` as \`b1__book_id\` from \`book_tag\` as \`b0\` left join \`book_tags\` as \`b1\` on \`b0\`.\`id\` = \`b1\`.\`book_tag_id\` where \`b2\`.\`book_id\` = 1",
  ],
]
`;

exports[`GH #5019 query 2 1`] = `
[
  [
    "[query] select \`b0\`.*, \`b1\`.\`id\` as \`b1__id\`, \`b1\`.\`title\` as \`b1__title\` from \`book_tag\` as \`b0\` left join \`book_tags\` as \`b2\` on \`b0\`.\`id\` = \`b2\`.\`book_tag_id\` left join \`book\` as \`b1\` on \`b2\`.\`book_id\` = \`b1\`.\`id\` where \`b3\`.\`book_id\` = 1",
  ],
]
`;

exports[`GH #5019 query 3 1`] = `
[
  [
    "[query] select \`b0\`.*, \`b1\`.\`id\` as \`b1__id\`, \`b1\`.\`title\` as \`b1__title\` from \`book_tag\` as \`b0\` left join \`book_tags\` as \`b2\` on \`b0\`.\`id\` = \`b2\`.\`book_tag_id\` left join \`book\` as \`b1\` on \`b2\`.\`book_id\` = \`b1\`.\`id\` where \`b0\`.\`id\` = 1",
  ],
]
`;
